### PR TITLE
Playground

### PR DESCRIPTION
**Pull request by brenkao**
- Added Playground - Click Open Playground in Traces Table.
    - You can create new versions
    - You can vibe check, which will also create new version if any changes are made
- Added Components for Call Params 
    - Added only mandatory call params
    - Optional call params have not been added yet
- Made major changes to db, this is a breaking change. Hopefully last breaking change.
- Fixed various UI bugs when viewing traces
- Added more error checking when writing prompts

### CODE REVIEW

File: lilypad/server/main.py [hunk 1]
```diff
@@ -208,6 +210,113 @@ async def get_llm_function_id_by_hash(
...
 
 
+def create_dynamic_function(
+    name: str, args_dict: dict[str, str], return_annotation: type
+):
+    """Create a dynamic function with the given name, arguments, and return type."""
+    arg_list = [f"{arg_name}: {arg_type}" for arg_name, arg_type in args_dict.items()]
+    arg_string = ", ".join(arg_list)
+
+    func_def = f"def {name}({arg_string}) -> {return_annotation.__name__}:\n    ..."
```
Comment (willbakst) [inline lilypad/server/main.py:220]: nit: does this need the `\n`?

Comment (brenkao) [reply lilypad/server/main.py:220]: oh, no it doesnt

File: lilypad/server/main.py [hunk 2]
```diff
@@ -206,6 +198,84 @@ async def get_llm_function_id_by_hash(
...
+    )
+
+    fn = create_dynamic_function(
+        name=version.llm_fn.function_name,
+        args_dict=args_dict,
+        return_annotation=str,
+    )
+
+    if not version.fn_params:
+        return trace_decorator(fn)(**arg_values)
```
Comment (brenkao) [inline lilypad/server/main.py:251]: should i just type ignore this or is there a way to get typing

File: lilypad/server/main.py [hunk 3]
```diff
@@ -206,6 +198,84 @@ async def get_llm_function_id_by_hash(
...
+
+    fn_params = FnParamsPublic.model_validate(version.fn_params)
+    return traced_synced_llm_function_constructor(fn_params, trace_decorator)(fn)(
```
Comment (brenkao) [inline lilypad/server/main.py:254]: ditto

File: lilypad/models/fn_params.py [hunk 4]
```diff
@@ -1,24 +1,63 @@
...
+class ResponseFormat(BaseModel):
+    """Response format model."""
+
+    type: Literal["text", "json_object", "json_schema"]
+
+
+class OpenAICallArgsCreate(BaseModel):
+    """OpenAI call args model."""
+
+    max_tokens: int
```
Comment (willbakst) [inline lilypad/models/fn_params.py:19]: not all of these are required fields?

File: lilypad/models/fn_params.py [hunk 5]
```diff
@@ -1,24 +1,63 @@
...
+    frequency_penalty: float
+    presence_penalty: float
+    response_format: ResponseFormat
+    stop: str | list[str] | None = None
+
+
+class AnthropicCallArgsCreate(BaseModel):
+    """Anthropic call args model."""
+
+    max_tokens: int
```
Comment (willbakst) [inline lilypad/models/fn_params.py:31]: same here

File: lilypad/cli/commands/start.py [hunk 6]
```diff
@@ -51,7 +51,7 @@ def start_lilypad(project_dir: Path, port: str) -> subprocess.Popen:
     env["LILYPAD_PROJECT_DIR"] = str(project_dir)
     files_dir = files("lilypad").joinpath("server")
     process = subprocess.Popen(
-        ["fastapi", "dev", "--port", port],
+        ["fastapi", "dev", "--port", str(port)],
```
Comment (willbakst) [inline lilypad/cli/commands/start.py:54]: there's another issue with `lilypad.configure` where it hardcodes port 8000 and instead needs to read from the config file, please fix that

File: lilypad/models/fn_params.py (hunk 4 above)
Comment (brenkao) [reply lilypad/models/fn_params.py:19]: they arent required but they have default values, so they are always set to default in the UI

File: lilypad/server/main.py (hunk 2 above)
Comment (willbakst) [reply lilypad/server/main.py:251]: just use the specific `pyright: ignore [{SPECIFIC_ERROR}]` for now, we can go through type ignores later and fix them, but not super high priority right now so long as it works

File: lilypad/server/main.py (hunk 3 above)
Comment (willbakst) [reply lilypad/server/main.py:254]: same here as my other comment

File: lilypad/cli/commands/start.py (hunk 6 above)
Comment (brenkao) [reply lilypad/cli/commands/start.py:54]:
```
    config = load_config()
    port = config.get("port", 8000)
    otlp_exporter = JSONSpanExporter(
        base_url=f"http://127.0.0.1:{port}/api",
    )
```
this is the code inside configure, are you saying its not pulling the port properly?

Comment (willbakst) [reply lilypad/cli/commands/start.py:54]: oh hmm, for some reason it wasn't tracing, not sure why

Comment (brenkao) [reply lilypad/cli/commands/start.py:55]: there was a bug in start that didnt set the config file properly, which may have caused a weird state.

File: lilypad/models/fn_params.py (hunk 4 above)
Comment (willbakst) [reply lilypad/models/fn_params.py:19]: but many of them can be null/none? the api doesn't require them

File: lilypad/models/fn_params.py (hunk 5 above)
Comment (willbakst) [reply lilypad/models/fn_params.py:31]: max_tokens is required, but not all others are

**Review by willbakst — APPROVED** [review]

**Pull request merged**